checker/src/main/java/org/checkerframework/checker/nullness/KeyForPropagationTreeAnnotator.java

```diff
@@ -102,4 +109,98 @@ public Void visitNewClass(NewClassTree tree, AnnotatedTypeMirror type) {
     keyForPropagator.propagateNewClassTree(tree, type, (KeyForAnnotatedTypeFactory) atypeFactory);
     return super.visitNewClass(tree, type);
   }
+
+  /**
+   * When visiting {@code Map.keySet()} calls, merge the map's key {@code @KeyFor} into the returned
+   * Set element.
+   *
+   * <p>{@inheritDoc}
+   */
+  @Override
+  public Void visitMethodInvocation(MethodInvocationTree tree, AnnotatedTypeMirror type) {
+    if (isCallToKeyset(tree) && type.getKind() == TypeKind.DECLARED) {
+      AnnotatedDeclaredType keySetReturnType = (AnnotatedDeclaredType) type;
+
+      // Get the receiver type
+      AnnotatedTypeMirror receiverType = atypeFactory.getReceiverType(tree);
+      if (receiverType != null
+          && receiverType.getKind() == TypeKind.DECLARED
+          && !keySetReturnType.getTypeArguments().isEmpty()) {
```

Comment:
If this is a call to `keySet()`, how could the type arguments be empty?  (Because of raw types??  In that case, please add a comment.)  Would it be better to write an assertion rather than putting a test in the `if`?

The same comment about test vs. assertion applies to `getKind() == DECLARED` immediately above.